File: twitstream/twitstream.cpp
#include <string>
#include <vector>
#include <iostream>
#include <curl/curl.h>
#include <json_spirit_reader_template.h>
#include <boost/program_options.hpp>
#include <boost/algorithm/string/join.hpp>


size_t callback(void * ptr, size_t size, size_t nmemb, void *)
{
  json_spirit::Value value;
  if(json_spirit::read_string(std::string(static_cast<const char *>(ptr), size * nmemb), value))
  {
    if(value.type() == json_spirit::obj_type)
    {
      for(const auto & item : value.get_obj())
      {
        if(item.name_ == "text" && item.value_.type() == json_spirit::str_type)
        {
          std::cout << item.value_.get_str() << std::endl;
        }
      }
    }
  }
  return size * nmemb;
}

namespace po = boost::program_options;

int main(int argc, char ** argv)
{
  std::string username;
  std::string password;
  std::vector<std::string> keywords;
  po::options_description desc("Options");
  desc.add_options()
    ("help,h", "display this help message")
    ("username,u", po::value<std::string>(&username), "twitter account user name")
    ("password,p", po::value<std::string>(&password), "twitter account password")
    ("keyword,k", po::value<std::vector<std::string> >(&keywords), "keywords to search");
  
  po::variables_map vm;
  po::store
    (po::parse_command_line(argc, argv, desc), 
     vm);
  po::notify(vm);
  
  if(vm.count("help"))
  {
    std::cout << desc << std::endl;
  }
  else
  {
    std::string track = std::string("track=") + boost::join(keywords, ",");
    
    CURL * curl = curl_easy_init();
    curl_easy_setopt(curl, CURLOPT_USERPWD, (username + ":" + password).c_str());
    curl_easy_setopt(curl, CURLOPT_URL, "https://stream.twitter.com/1/statuses/filter.json");
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, callback);
    curl_easy_setopt(curl, CURLOPT_POST, 1);
    curl_easy_setopt(curl, CURLOPT_POSTFIELDS, track.data());
    curl_easy_setopt(curl, CURLOPT_POSTFIELDSIZE, track.size());
    curl_easy_perform(curl);
    
    curl_easy_cleanup(curl);
  }
}

File: mltwitstream/twitstream.ml
let callback e = 
  (
    try
      let json = Yojson.Safe.from_string e in
      match json with
          `Assoc elements -> 
            (
              match (List.assoc "text" elements) with
                  `String e -> prerr_endline e
                | _ -> raise Not_found
            )
        | _ -> raise Not_found
    with _ -> prerr_endline "Error reading message"
  );
  String.length e
    
let _ =
  let username = ref ""
  and password = ref ""
  and keywords = ref [] in
  Arg.parse 
    ["-u", Arg.Set_string username, "username";
     "-p", Arg.Set_string password, "password"]
    (fun e -> keywords := e::(!keywords))
    "twitstream [flags] [keywords]";
  
  let all_keywords = "track=" ^ (String.concat ":" (!keywords)) in
  let c = Curl.init() in
  Curl.setopt c (Curl.CURLOPT_USERPWD (!username ^ ":" ^ !password));
  Curl.setopt c (Curl.CURLOPT_URL "https://stream.twitter.com/1/statuses/filter.json");
  Curl.setopt c (Curl.CURLOPT_WRITEFUNCTION callback);
  Curl.setopt c (Curl.CURLOPT_POST true);
  Curl.setopt c (Curl.CURLOPT_POSTFIELDS all_keywords);
  Curl.setopt c (Curl.CURLOPT_POSTFIELDSIZE (String.length all_keywords));
  Curl.perform c;
  Curl.cleanup c
    
